### PR TITLE
Build System for 4.0

### PR DESCRIPTION
**Pull request by varon**
### Purpose of this PR

Adds build automation with the intention of moving to [continuous delivery](https://continuousdelivery.com/).

Four build systems were reviewed prior to this choice:
- [Bullseye](https://github.com/adamralph/bullseye)
- [CAKE](https://cakebuild.net/)
- [FAKE](https://fake.build/)
- [Nuke](https://nuke.build/)

### Why FAKE?
- No external or global tool required install to edit or run the scripts.
- Full IDE support without any tooling.
- Excellent bootstrapping process.
- Maturity and simplicity.
- Ready availability of other projects as examples.

### Current Status

- [x] Bootstrapping
- [x] Configuration
- [x] Package Restore
- [x] Cleaning
- [ ] AssemblyInfo/Property updates
- [x] Building release projects
- [x] Building tooling (internal) projects
- [x] Building tests
- [x] Running tests
- [x] Update specification task

- [x] Update Bindings task
- [x] Build NuGet Packages
- [x] Release on Github
- [x] Release on Nuget.Org

#### Targets for bindings/specification updates

I'd love to receive PRs to update the specification and re-generate bindings into build tasks. [This is available for 3.0](https://github.com/opentk/opentk/blob/3.0/build.fsx#L168), and 4.0 on Windows via two batch scripts ([update-bindings.bat](https://github.com/opentk/opentk/blob/master/update-bindings.bat) and [update-specifications.bat](https://github.com/opentk/opentk/blob/master/update-specifications.bat)).

These should be added into the build script as simple, cross-platform targets.

#### Target to run the unit tests

We need to run the unit tests and check they pass. Some code is included, but mostly it's fiddling with dependencies to make sure that we're getting the testrunner as part of the bootstrapping process.

#### Target to create the NuGet packages

We probably need to define template or metadata information and provide this to the NuGet module in FAKE. The best approach we want to take for this open to discussion.

#### Create a target to update version info

We want to be able to automatically update the version info for the project, based on what we've written in RELEASE_NOTES.md. FAKE has built-in mechanisms to parse them, we just need to figure out the best approach to use this in the NuGet package metadata.

At the moment we're manually including a bunch of `.props` files. Perhaps using the recent implicit property inclusion via [Directory.Build.props](https://docs.microsoft.com/en-us/visualstudio/msbuild/customize-your-build?view=vs-2019#directorybuildprops-and-directorybuildtargets) may offer us a simpler and cleaner approach.

Whichever we go with, we need to make sure that we have an easy mechanism to go from RELEASE_NOTES.md to our release metadata.

#### General reviews/comments

Happy to answer any questions, and feedback and comments are absolutely welcome on this!

### CODE REVIEW

**Comment by realvictorprm** [comment]
You can tick `BuildingTests` so far I tested and looked through your build script.

**Comment by varon** [comment]
Thanks a ton for all the help so far on this @realvictorprm .

Updated and re-ordered the checklist as we chatted about.

File: build.fsx [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+
+Target.create "Restore" (fun _ ->
+     Shell.Exec("dotnet", "restore OpenTK.sln") |> ignore
+ )
+
+// Generate assembly info files with the right version & up-to-date information
+Target.create "AssemblyInfo" (fun _ ->
+    //TODO: Create and update the Directory.Build.props file with the version information taken from the releaseNotes value.
+    // see https://docs.microsoft.com/en-us/visualstudio/msbuild/customize-your-build?view=vs-2019#directorybuildprops-and-directorybuildtargets
+    Trace.traceError "Unimplemented."
```
Comment (varon) [inline build.fsx]: still NYI here.

Comment (varon) [inline build.fsx]: Spoke to victor. He'll tackle that a bit later on.

**Comment by varon** [comment]
@realvictorprm Needs the WIP tag removed prior to merge.

**Comment by realvictorprm** [comment]
@varon you can remove it

**Comment by matkoch** [comment]
Hey all :) I'm late to the party, but I'm curious which requirements nuke didn't meet.

**Comment by varon** [comment]
@matkoch  - very cool to see you guys actively hunting issues and comparisons!

For the record, Nuke was a really close, strong second. it's a great build system and a much more sane design than CAKE. Especially like the symbolic rather than string-based targets.

From what I can recall (this was a while ago) there were two major issues with NUKE at the time:
1) The bootstrapping process wasn't great (requires a global tool install?)
2) It looks like we need IDE extensions to get full functionality.

I think finally, when I did try to include it, the samples didn't bootstrap as cleanly as FAKE did, especially as in our case, the solution will not compile until we've performed our code generation step (which from vague memory stopped the build from executing).

One final factor we had to consider is that all of our old build system was written in FAKE as well. This works extremely well for OpenTK 3.x, so working from that greatly reduced the hours required to get everything up and running.

It's really great to see you out here hunting and trying to figure out why we didn't adopt your library (this time, for now). You've set a great example as a maintainer, and I look forward to trying Nuke in some other projects down the road!

Hope that helps,
Varon

**Comment by matkoch** [comment]
Thanks for the quick reply :) Not to convince you changing your opinion now, but here is my answer, and maybe you have any advice what part exactly wasn't clear.

1. Bootstrapping is done via `build.ps1` and `build.sh`. Both scripts will take care of installing the proper (`global.json`) .NET Core SDK. If one exists locally, it will happily use that. Documentation is [here](http://www.nuke.build/docs/running-builds/fundamentals.html). What you're talking about in regards to the global tool is for the [setup](http://www.nuke.build/docs/getting-started/setup.html). That allows to include the build project by answering a few questions. In addition, it also offers [shell completion](http://www.nuke.build/docs/running-builds/global-tool.html) for the actual build.

2. No IDE extensions are required, since the build is implemented in a simple console application. So you could just mark the project as startup project, and it will just execute. However, there are a few [IDE extensions](http://www.nuke.build/docs/running-builds/from-ides.html) for enhanced support, which means you can execute/debug targets with and without dependencies.

I'm not sure how far you got using NUKE, but the part about code generation sounds like you've added it manually to the solution. At least from the description, I don't see any issue why this shouldn't work.

Alright, that was basically just for reference, and I appreciate your quick reply very much! :)

**Pull request merged**